### DISCUSSION TITLE
请教下，RichEdit 控件，如何获取内容的行数

### DISCUSSION THREAD

**Discussion by huangzhengyi** [netease-im/NIM_Duilib_Framework#369] [General]
![image](https://user-images.githubusercontent.com/23769812/144561627-3cab2381-8498-4edd-93c2-e2b0cd89e540.png)
![image](https://user-images.githubusercontent.com/23769812/144561736-c7e25e1c-054d-48f0-91d4-67a3ee64a50d.png)
通过 GetLineCount 获取行数，由于m_pTwh 还没有初始化，这种情况应该怎么去实现

**Comment by huangzhengyi** [comment]
有人知道怎么实现嘛？怎么获取行数

**Comment by nmgwddj** [comment]
该变量会在 DoInit 中进行初始化，你应该关心的是 DoInit 后为什么 m_pTwh 为空，断点到 DoInit 中查看一下关键代码段的返回值相信可以找到你的问题根源：
https://github.com/netease-im/NIM_Duilib_Framework/blob/master/duilib/Control/RichEdit.cpp#L2133

**Reply by huangzhengyi** [reply]
new RichEdit()  之后调用 GetLineCount  会比 DoInit  优先调用，不知道是为什么

**Reply by nmgwddj** [reply]
在 GetLineCount 下断点，查看调用栈是从什么位置过来的。

**Reply by huangzhengyi** [reply]
我是在 InitWindow() 的时候 new 这个控件的，然后马上调用GetLineCount

**Reply by huangzhengyi** [reply]
试了下 InitWindow 之后，通过点击按钮来new 然后再调用，也是这个问题

**Reply by nmgwddj** [reply]
new 出来的 RichEdit 立即调用 GetLineCount 的意义在哪里呢？

**Reply by huangzhengyi** [reply]
我是动态的创建一个edit，然后根据有多少行来设置这个控件的高度

**Reply by nmgwddj** [reply]
可以将你的示例代码传到自己的 github 仓库中，我尝试克隆下来测试。

**Reply by huangzhengyi** [reply]
![image](https://user-images.githubusercontent.com/23769812/144961931-0625d721-7e5b-4301-9549-d4450c9cd13c.png)

**Comment by nmgwddj** [comment]
GetLineCount 要在 m_pTwh 初始化后才会返回正确值。
```diff
int RichEdit::GetLineCount() const
{ 
+    if( !m_pTwh ) return 0;
    LRESULT lResult;
    TxSendMessage(EM_GETLINECOUNT, 0, 0, &lResult);
    return (int)lResult; 
}
```
m_pTwh 的初始化在 DoInit 接口中
```diff
void RichEdit::DoInit()
{
	if (m_bInited)
		return;

	CREATESTRUCT cs;
	cs.style = m_lTwhStyle;
	cs.x = 0;
	cs.y = 0;
	cs.cy = 0;
	cs.cx = 0;
	cs.lpszName = m_sText.c_str();
+	CreateHost(this, &cs, &m_pTwh);
	if (m_pTwh) {
		m_pTwh->SetTransparent(TRUE);
		LRESULT lResult;
		m_pTwh->GetTextServices()->TxSendMessage(EM_SETLANGOPTIONS, 0, 0, &lResult);
		m_pTwh->GetTextServices()->TxSendMessage(EM_SETEVENTMASK, 0, ENM_CHANGE, &lResult);
		m_pTwh->OnTxInPlaceActivate(NULL);
		if (m_pHorizontalScrollBar) {
			m_pHorizontalScrollBar->SetScrollRange(0);
		}
		if (m_pVerticalScrollBar) {
			m_pVerticalScrollBar->SetScrollRange(0);
		}
	}

	m_bInited = true;
}
```
DoInit 接口在基类 PlaceHolder 的 Init 接口中会调用
```diff
void PlaceHolder::Init()
{
+	DoInit();
}
```
PlaceHolder::Init 方法在 SetWindow 方法中会调用
```diff
void PlaceHolder::SetWindow(Window* pManager, Box* pParent, bool bInit)
{
	m_pWindow = pManager;
	m_pParent = pParent;
+	if (bInit && m_pParent) Init();
}
```
你 new 了新的控件，给他设置了各种样式，但是你没有告诉它应该显示在什么位置。

**Reply by huangzhengyi** [reply]
非常感谢，解决了。
需要先将edit 添加到 list，再去设置样式之后再获取 ，就可以了

**Reply by nmgwddj** [reply]
Good，另外给您一些反馈问题的建议。
代码托管平台重要的是代码而不是图片，反馈问题时使用 markdown 格式贴关键位置代码，善用 markdown 语法标签可以让你的问题表达更加明确，减少没必要的循环排查。站在接收问题人的角度考虑，其他人看到你反馈的问题和相应的代码能否快速介入协助你排查问题。
